be/src/olap/hll.h

```diff
@@ -33,14 +33,339 @@ const static int HLL_REGISTERS_COUNT = 16384;
+    typedef int32_t SparseLengthValueType;
+    typedef uint16_t SparseIndexType;
+    typedef uint8_t SparseValueType;
+
+    static void update_registers(char* registers, uint64_t hash_value) {
+        // Use the lower bits to index into the number of streams and then
+        // find the first 1 bit after the index bits.
+        int idx = hash_value % HLL_REGISTERS_COUNT;
+        uint8_t first_one_bit = __builtin_ctzl(hash_value >> HLL_COLUMN_PRECISION) + 1;
+        registers[idx] = std::max((uint8_t)registers[idx], first_one_bit);
+    }
+
+    static void merge_hash_set_to_registers(char* registers, const std::set<uint64_t>& hash_set) {
+        for (auto hash_value: hash_set) {
+            update_registers(registers, hash_value);
+        }
+    }
+
+    static void merge_registers(char* registers, const char* other_registers) {
+        for (int i = 0; i < doris::HLL_REGISTERS_COUNT; ++i) {
+            registers[i] = std::max(registers[i], other_registers[i]);
+        }
+    }
+
+    static int serialize_full(char* result, char* registers) {
+        result[0] = HLL_DATA_FULL;
+        memcpy(result + 1, registers, HLL_REGISTERS_COUNT);
+        return HLL_COLUMN_DEFAULT_LEN;
+    }
+
+    static int serialize_sparse(char *result, const std::map<int, uint8_t>& index_to_value) {
+        result[0] = HLL_DATA_SPRASE;
+        int len = sizeof(SetTypeValueType) + sizeof(SparseLengthValueType);
+        char* write_value_pos = result + len;
+        for (auto iter = index_to_value.begin();
+             iter != index_to_value.end(); iter++) {
+            write_value_pos[0] = (char)(iter->first & 0xff);
+            write_value_pos[1] = (char)(iter->first >> 8 & 0xff);
+            write_value_pos[2] = iter->second;
+            write_value_pos += 3;
+        }
+        int registers_count = index_to_value.size();
+        len += registers_count * (sizeof(SparseIndexType)+ sizeof(SparseValueType));
+        *(int*)(result + 1) = registers_count;
+        return len;
+    }
+
+    static int serialize_explicit(char* result, const std::set<uint64_t>& hash_value_set) {
+        result[0] = HLL_DATA_EXPLICIT;
+        result[1] = (uint8_t)(hash_value_set.size());
+        int len = sizeof(SetTypeValueType) + sizeof(uint8_t);
+        char* write_pos = result + len;
+        for (auto iter = hash_value_set.begin();
+             iter != hash_value_set.end(); iter++) {
+            uint64_t hash_value = *iter;
+            *(uint64_t*)write_pos = hash_value;
+            write_pos += 8;
+        }
+        len += sizeof(uint64_t) * hash_value_set.size();
+        return len;
+    }
+
+    // change the _type to HLL_DATA_FULL directly has two reasons:
+    // 1. keep behavior consistent with before
+    // 2. make the code logic is simple
+    void update(const uint64_t hash_value) {
+        _type = HLL_DATA_FULL;
+        update_registers(_registers, hash_value);
+    }
+
+    void merge(const HyperLogLog& other) {
```

Comment:
For large code block, we'd better to write it in cpp/cc files to accelerate our compile time

Reply:
OK